skimage/morphology/_util.py

```diff
@@ -51,35 +54,188 @@ def _validate_connectivity(image_dim, connectivity, offset):
         The shape of the image for which the offsets are computed.
-    structure : ndarray
+    selem : ndarray
         A structuring element determining the neighborhood expressed as an
         n-D array of 1's and 0's.
-    center : sequence
-        Tuple of indices specifying the center of `selem`.
+    center : tuple
+        Tuple of indices to the center of `selem`.
+    order : {"C", "F"}, optional
+        Whether the image described by `image_shape` is in row-major (C-style)
+        or column-major (Fortran-style) order.
 
     Returns
     -------
-    offsets : ndarray
+    raveled_offsets : ndarray
         Linear offsets to a samples neighbors in the raveled image, sorted by
-        their Euclidean distance from the center.
+        their distance from the center.
+
+    Notes
+    -----
+    This function will return values even if `image_shape` contains a dimension
+    length that is smaller than `selem`.
 
     Examples
     --------
     >>> _offsets_to_raveled_neighbors((4, 5), np.ones((4, 3)), (1, 1))
     array([-5, -1,  1,  5, -6, -4,  4,  6, 10,  9, 11])
     """
-    structure = structure.copy()  # Don't modify original input
-    structure[tuple(center)] = 0  # Ignore the center; it's not a neighbor
-    connection_indices = np.transpose(np.nonzero(structure))
-    offsets = (np.ravel_multi_index(connection_indices.T, image_shape,
-                                    order=order) -
-               np.ravel_multi_index(center, image_shape, order=order))
-    squared_distances = np.sum((connection_indices - center) ** 2, axis=1)
-    return offsets[np.argsort(squared_distances)]
+    if not selem.ndim == len(image_shape) == len(center):
+        raise ValueError(
+            "number of dimensions in image_shape, selem and center "
+            "does not match"
+        )
+
+    selem_indices = np.array(np.nonzero(selem)).T
+    offsets = selem_indices - center
+
+    if order == 'F':
+        offsets = offsets[:, ::-1]
+        image_shape = image_shape[::-1]
+    elif not order == 'C':
+        raise ValueError("order was not 'C' or 'F'")
+
+    # Scale offsets in each dimension and sum
+    ravel_factors = image_shape[1:] + (1,)
+    ravel_factors = np.cumprod(ravel_factors[::-1])[::-1]
+    raveled_offsets = (offsets * ravel_factors).sum(axis=1)
+
+    # Sort by distance
+    distances = np.abs(offsets).sum(axis=1)
+    raveled_offsets = raveled_offsets[np.argsort(distances)]
+
+    # In case any dimension in image_shape is smaller than selem.shape
+    # duplicates might occur, remove them
+    if any(x < y for x, y in zip(image_shape, selem.shape)):
+        # np.unique reorders, which we don't want
+        _, indices = np.unique(raveled_offsets, return_index=True)
+        raveled_offsets = raveled_offsets[np.sort(indices)]
+
+    # Remove "offset to center"
+    raveled_offsets = raveled_offsets[1:]
+
+    return raveled_offsets
+
+
+def _resolve_neighborhood(selem, connectivity, ndim):
```

Comment:
If I'm correct, this function has no doctest nor unitest. That would deserve to open an issue.

Reply:
All of the functions in this module are only implicitly tested through tests for the public functions. I'm hesitant to add detailed tests here because these are just convenience functions (excluding `_offsets_to_raveled_neighbors`) dealing with input and I'd rather test the public API than parts of the implementation that may change significantly down the line. E.g. `_fast_pad` may be safely removed once NumPy 1.17 is given by scikit-image's dependencies.

However `_offsets_to_raveled_neighbors` could indeed profit from a few explicit tests. I'll try to come up with a few meaningful tests...

Reply:
Done.